File: task3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <numeric>
#include <cstdint>

using namespace std;

// НОД для беззнаковых чисел
uint64_t gcd(uint64_t a, uint64_t b) {
    while (b != 0) {
        uint64_t t = b;
        b = a % b;
        a = t;
    }
    return a;
}

// Рекуррентное вычисление коэффициентов Эйлера
// Возвращает вектор E(a, k) для k = 0..a-1
vector<uint64_t> eulerCoeff(int a) {
    if (a < 1) return {};
    // Треугольник Эйлера: строки от 1 до a
    vector<vector<uint64_t>> triangle(a + 1);
    
    for (int n = 1; n <= a; n++) {
        triangle[n].resize(n);
        triangle[n][0] = 1;           // первый элемент всегда 1
        triangle[n][n - 1] = 1;       // последний элемент всегда 1
        
        for (int k = 1; k < n - 1; k++) {
            // E(n, k) = (n - k) * E(n-1, k-1) + (k + 1) * E(n-1, k)
            triangle[n][k] = (n - k) * triangle[n - 1][k - 1] 
                           + (k + 1) * triangle[n - 1][k];
        }
    }
    
    return triangle[a];
}

// Функция сокращения дроби
void reduce(uint64_t& num, uint64_t& den) {
    uint64_t g = gcd(num, den);
    num /= g;
    den /= g;
}

// Вычисление суммы ряда
string computeSum(int a, int b) {
    if (b == 1) {
        return "infinity";
    }
    
    uint64_t bb = b;
    
    // Получаем коэффициенты Эйлера для данного a
    vector<uint64_t> coeff = eulerCoeff(a);
    
    // Вычисляем A_a(b) = sum_{k=0}^{a-1} coeff[k] * b^k
    uint64_t A = 0;
    uint64_t power = 1;  // b^0
    for (int k = 0; k < a; k++) {
        A += coeff[k] * power;
        power *= bb;  // переходим к следующей степени b
    }
    
    uint64_t numerator = bb * A;
    uint64_t denominator = 1;
    for (int i = 0; i < a + 1; i++) {
        denominator *= (bb - 1);
    }
    
    reduce(numerator, denominator);
    
    return to_string(numerator) + "/" + to_string(denominator);
}

int main() {
    int a, b;
    cin >> a >> b;
    cout << computeSum(a, b) << endl;
    return 0;
}

File: rust/src/task3.rs
use std::io;

fn gcd(mut a: i64, mut b: i64) -> i64 {
    while b != 0 {
        let t = b;
        b = a % b;
        a = t;
    }
    a.abs()
}

fn reduce(num: &mut i64, den: &mut i64) {
    let g = gcd(*num, *den);
    *num /= g;
    *den /= g;
}

fn euler_coeff(a: usize) -> Vec<i64> {
    if a == 0 {
        return vec![];
    }
    
    let mut triangle: Vec<Vec<i64>> = vec![vec![]; a + 1];
    
    for n in 1..=a {
        triangle[n] = vec![0; n];
        triangle[n][0] = 1;
        triangle[n][n - 1] = 1;
        
        for k in 1..n - 1 {
            triangle[n][k] = (n - k) as i64 * triangle[n - 1][k - 1]
                           + (k + 1) as i64 * triangle[n - 1][k];
        }
    }
    
    triangle[a].clone()
}

fn compute_sum(a: i32, b: i32) -> String {
    if b == 1 {
        return "infinity".to_string();
    }
    
    let a_usize = a as usize;
    let bb = b as i64;
    
    let coeff = euler_coeff(a_usize);
    
    let mut a = 0i64;
    let mut power = 1i64;
    for k in 0..a_usize {
        a += coeff[k] * power;
        power *= bb;
    }
    
    let numerator = bb * a;
    
    let mut denominator = 1i64;
    for _ in 0..(a + 1) {
        denominator *= bb - 1;
    }
    
    let mut num = numerator;
    let mut den = denominator;
    reduce(&mut num, &mut den);
    
    format!("{}/{}", num, den)
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let parts: Vec<i32> = input
        .split_whitespace()
        .map(|x| x.parse().expect("Not a number"))
        .collect();
    let (a, b) = (parts[0], parts[1]);
    println!("{}", compute_sum(a, b));
}